tools/rpkg/tests/testthat/test_arrow.R

```diff
@@ -0,0 +1,295 @@
+
+  # And we can continue the pipeline
+  ds_rt <- ds %>%
+    to_duckdb() %>%
+    # factors don't roundtrip https://github.com/duckdb/duckdb/issues/1879
+    select(-fct) %>%
+    to_arrow() %>%
+    filter(int > 5)
+
+  expect_identical(
+    ds_rt %>%
+      collect() %>%
+      arrange(int),
+    ds %>%
+      select(-fct) %>%
+      filter(int > 5) %>%
+      collect() %>%
+      arrange(int)
+  )
+
+  # Now check errors
+  ds_rt <- ds %>%
+    to_duckdb() %>%
+    # factors don't roundtrip https://github.com/duckdb/duckdb/issues/1879
+    select(-fct)
+
+  # alter the class of ds_rt's connection to simulate some other database
+  class(ds_rt$src$con) <- "some_other_connection"
+
+  expect_error(
+    to_arrow(ds_rt),
+    "to_arrow\\(\\) currently only supports Arrow tables, Arrow datasets,"
+  )
+})
+
+test_that("to_arrow roundtrip, with dataset", {
+  # these will continue to error until 0.3.2 is released
+  # https://github.com/duckdb/duckdb/pull/2957
+  skip_if_not_installed("duckdb", minimum_version = "0.3.2")
+  # With a multi-part dataset
+  tf <- tempfile()
+  new_ds <- rbind(
+    cbind(example_data, part = 1),
+    cbind(example_data, part = 2),
+    cbind(mutate(example_data, dbl = dbl * 3, dbl2 = dbl2 * 3), part = 3),
+    cbind(mutate(example_data, dbl = dbl * 4, dbl2 = dbl2 * 4), part = 4)
+  )
+  write_dataset(new_ds, tf, partitioning = "part")
+
+  ds <- open_dataset(tf)
+
+  expect_identical(
+    ds %>%
+      to_duckdb() %>%
+      select(-fct) %>%
+      mutate(dbl_plus = dbl + 1) %>%
+      to_arrow() %>%
+      filter(int > 5 & part > 1) %>%
+      collect() %>%
+      arrange(part, int) %>%
+      as.data.frame(),
+    ds %>%
+      select(-fct) %>%
+      filter(int > 5 & part > 1) %>%
+      mutate(dbl_plus = dbl + 1) %>%
+      collect() %>%
+      arrange(part, int)
+  )
+})
+
+# test_that("to_arrow roundtrip, with dataset (without wrapping", {
+#   # these will continue to error until 0.3.2 is released
+#   # https://github.com/duckdb/duckdb/pull/2957
+#   skip_if_not_installed("duckdb", minimum_version = "0.3.2")
+#   # With a multi-part dataset
+#   tf <- tempfile()
+#   new_ds <- rbind(
+#     cbind(example_data, part = 1),
+#     cbind(example_data, part = 2),
+#     cbind(mutate(example_data, dbl = dbl * 3, dbl2 = dbl2 * 3), part = 3),
+#     cbind(mutate(example_data, dbl = dbl * 4, dbl2 = dbl2 * 4), part = 4)
+#   )
+#   write_dataset(new_ds, tf, partitioning = "part")
+
+#   out <- ds %>%
+#     to_duckdb() %>%
+#     select(-fct) %>%
+#     mutate(dbl_plus = dbl + 1) %>%
+#     to_arrow(as_arrow_query = FALSE)
+
+#   expect_r6_class(out, "RecordBatchReader")
+# })
+
+# The next set of tests use an already-extant connection to test features of
+# persistence and querying against the table without using the `tbl` itself, so
+# we need to create a connection separate from the ephemeral one that is made
+# with arrow_duck_connection()
+con <- dbConnect(duckdb::duckdb())
+dbExecute(con, "PRAGMA threads=2")
```

Comment:
These appear to be outside `test_that()`-function. Maybe that's not the best approach? Connection is not explicitly closed here and most likely `on.exit()` will not be triggered here as it is not inside of a function and results in garbage collected connections?